src/librustc_codegen_llvm/mir/analyze.rs

```diff
@@ -152,44 +154,46 @@ impl Visitor<'tcx> for LocalAnalyzer<'mir, 'a, 'll, 'tcx> {
 
     fn visit_place(&mut self,
                    place: &mir::Place<'tcx>,
-                   context: PlaceContext<'tcx>,
+                   mut context: PlaceContext<'tcx>,
                    location: Location) {
         debug!("visit_place(place={:?}, context={:?})", place, context);
         let cx = self.fx.cx;
+        let neo_place = cx.tcx.as_new_place(place);
+
+        let mut base_ty = neo_place.base.ty(self.fx.mir);
 
-        if let mir::Place::Projection(ref proj) = *place {
+        for elem in neo_place.elems.iter().rev() {
```

Comment:
So now you are going backwards, but I think that `base_ty` isn't really correct now =) -- that is, if you have a path like `a.b.c`, `base_ty` will start out as the type of `a`, but then you look at `c` first -- you need the type of `a.b` for that. 

Maybe it's a good idea to first construct a vector with the type at each projection -- where 0 is the type of the base.

Something like:

```rust
let mut base_ty = PlaceTy::from(neo_place.base.ty(self.fx.mir));
let mut place_tys: Vec<PlaceTy<'tcx>> = vec![base_ty];
for elem in neo_place.elems.iter() {
    let next_ty = place_ty.last().unwrap().projection_ty(tcx, elem);
    place_tys.push(next_ty);
    base_ty = next_ty;
}
```

then you can iterate backwards over the elements like so:

```rust
for elem in neo_place.elems.iter().zip(&place_tys[1..]).rev() {
    ...
}
```